src/Domain/LeanCode.DomainModels.MassTransitRelay/Testing/AsyncManualResetEvent.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using GreenPipes.Internals.Extensions;
+
+namespace LeanCode.DomainModels.MassTransitRelay.Testing
+{
+    // Based on https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncManualResetEvent.cs
+    public sealed class AsyncManualResetEvent
+    {
+        private readonly object mutex;
+        private TaskCompletionSource<ValueTuple> tcs;
+
+        public AsyncManualResetEvent(bool set)
+        {
+            mutex = new object();
+            tcs = new TaskCompletionSource<ValueTuple>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            if (set)
+            {
+                tcs.TrySetResult(default);
+            }
+        }
+
+        public AsyncManualResetEvent()
+            : this(false)
+        { }
+
+        public bool IsSet
+        {
+            get
+            {
+                lock (mutex)
+                {
+                    return tcs.Task.IsCompleted;
+                }
+            }
+        }
+
+        public Task WaitAsync()
+        {
+            lock (mutex)
+            {
+                return tcs.Task;
+            }
+        }
+
+        public Task WaitAsync(CancellationToken cancellationToken)
+        {
+            var waitTask = WaitAsync();
+            if (waitTask.IsCompleted)
+            {
+                return waitTask;
+            }
+            else
+            {
+                return waitTask.OrCanceled(cancellationToken);
+            }
+        }
+
+        public async Task<bool> WaitAsync(TimeSpan timeout)
+        {
+            try
+            {
+                await WaitAsync().OrTimeout(timeout);
+                return true;
+            }
+            catch (TimeoutException)
+            {
+                return false;
+            }
+        }
+
+        public void Set()
+        {
+            lock (mutex)
+            {
+                tcs.TrySetResult(default);
+            }
+        }
+
+        public void Reset()
+        {
+            lock (mutex)
+            {
+                if (tcs.Task.IsCompleted)
+                {
+                    tcs = new TaskCompletionSource<ValueTuple>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
One more thing, I just remembered; target typing should work in such cases: `tsc = new(TaskCreationOptions.RunContinuationsAsynchronously)`.

But I'm not sure if we want to use it yet.

Reply:
Ha, right. Let's check how does that look like. :)

Reply:
With a slight update to StyleCop - it works and looks quite nice. :)

Reply:
Also it appears that this release of StyleCop is aware of positional records, so it won't be angry with `record A(T Thing);` anymore :tada:

I shall fix my `SuppressMessage`s soon.